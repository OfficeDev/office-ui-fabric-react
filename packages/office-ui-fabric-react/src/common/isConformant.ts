// eslint-disable-next-line import/no-extraneous-dependencies -- this file is for testing
import { isConformant as isConformantBase, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: IsConformantOptions) {
  const defaultOptions = {
    disabledTests: ['has-docblock'],
  };

  isConformantBase(defaultOptions, testInfo);
}
