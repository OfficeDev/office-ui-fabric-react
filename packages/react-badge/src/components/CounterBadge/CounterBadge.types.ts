import * as React from 'react';
import { ComponentProps, ObjectShorthandProps } from '@fluentui/react-utilities';
import { BadgeProps } from '../Badge/index';

/**
 * {@docCategory CounterBadge}
 */
export type CounterBadgeAppearance = 'filled' | 'ghost';

/**
 * {@docCategory CounterBadge}
 */
export type CounterBadgeShape = 'rounded' | 'circular';

/**
 * {@docCategory CounterBadge}
 */
export type CounterBadgeColors = 'accent' | 'warning' | 'important' | 'severe' | 'informative';

/**
 * {@docCategory CounterBadge}
 */
export interface CounterBadgeProps
  extends ComponentProps,
    React.HTMLAttributes<HTMLElement>,
    Omit<BadgeProps, 'appearance' | 'shape'> {
  /**
   * A Badge can be square, circular or rounded
   * @defaultvalue circular
   */
  shape?: CounterBadgeShape;

  /**
   * A Badge can be filled, outline, ghost, inverted
   * @defaultvalue filled
   */
  appearance?: CounterBadgeAppearance;

  /**
   * A Badge can have color variations
   * @defaultvalue accent
   */
  color?: CounterBadgeColors;

  /**
   * Max number to be displayed
   * @defaultvalue 99
   */
  overflowCount?: number;

  /**
   * Value diplayed by the Badge
   * @defaultvalue 0
   */
  count?: number;

  /**
   * If 0 number should be displayed
   * @defaultvalue true
   */
  showZero?: boolean;
}

/**
 * {@docCategory CounterBadge}
 */
export interface CounterBadgeState extends CounterBadgeProps {
  /**
   * Ref to the root slot
   */
  ref: React.MutableRefObject<HTMLElement>;
  /**
   * Icon slot when processed by internal state
   */
  icon?: ObjectShorthandProps<HTMLSpanElement>;

  /**
   * Max number to be displayed
   * @defaultvalue 99
   */
  overflowCount: number;

  /**
   * Value diplayed by the Badge
   * @defaultvalue 0
   */
  count: number;
}
