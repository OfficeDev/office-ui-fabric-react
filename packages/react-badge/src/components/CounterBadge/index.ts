export * from './CounterBadge';
export * from './CounterBadge.types';
export * from './renderCounterBadge';
export * from './useCounterBadge';
export * from './useCounterBadgeStyles';
