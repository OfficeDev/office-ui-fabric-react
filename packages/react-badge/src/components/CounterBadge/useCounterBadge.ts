import * as React from 'react';
import { makeMergeProps, resolveShorthandProps } from '@fluentui/react-utilities';
import { CounterBadgeProps, CounterBadgeState } from './CounterBadge.types';
import { useBadge } from '../Badge/index';

/**
 * Consts listing which props are shorthand props.
 */
export const counterBadgeShorthandProps: string[] = ['icon'];

const mergeProps = makeMergeProps<CounterBadgeState>({ deepMerge: counterBadgeShorthandProps });

/**
 * Returns the props and state required to render the component
 */
export const useCounterBadge = (
  props: CounterBadgeProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: CounterBadgeProps,
): CounterBadgeState => {
  const count = 0;
  const overflowCount = 99;
  const state = mergeProps(
    useBadge(props, ref),
    {
      showZero: false,
      overflowCount,
      count,
      dot: false,
      ...(!props.dot && {
        children:
          (props.count || count) > (props.overflowCount || overflowCount) ? `${props.count}+` : `${props.count}`,
      }),
    },

    defaultProps,
    resolveShorthandProps(props, counterBadgeShorthandProps),
  );

  return state;
};
