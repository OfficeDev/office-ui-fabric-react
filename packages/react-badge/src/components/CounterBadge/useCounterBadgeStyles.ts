import { makeStyles, ax } from '@fluentui/react-make-styles';
import { CounterBadgeState } from './CounterBadge.types';
import { useRootStyles, useIconStyles } from '../Badge/useBadgeStyles';

/**
 * Styles for the root slot
 */
export const useCounterBadgeRootStyles = makeStyles<CounterBadgeState>([
  [
    null,
    {
      minWidth: 'auto',
    },
  ],
  [
    s => s.color === 'warning',
    theme => ({
      backgroundColor: theme.global.palette.cranberry.primary,
      borderColor: theme.global.palette.cranberry.primary,
    }),
  ],
  [
    s => s.color === 'important',
    theme => ({
      backgroundColor: theme.global.palette.grey[14],
      borderColor: theme.global.palette.grey[14],
    }),
  ],
  [
    s => s.color === 'severe',
    theme => ({
      // @TODO: update these colors once the color used in the design spec has existing color token
      backgroundColor: theme.global.palette.red.primary,
      borderColor: theme.global.palette.red.primary,
    }),
  ],
  [
    s => s.color === 'informative',
    theme => ({
      backgroundColor: theme.global.palette.grey[92],
      borderColor: theme.global.palette.grey[92],
      color: theme.alias.color.neutral.neutralForeground3,
    }),
  ],
  [
    s => s.count === 0 && !s.showZero,
    {
      width: '6px',
      height: '6px',
      padding: '0',
    },
  ],
]);

/**
 * Applies style classnames to slots
 */
export const useCounterBadgeStyles = (state: CounterBadgeState) => {
  state.className = ax(useRootStyles(state), useCounterBadgeRootStyles(state), state.className);
  const iconClassName = useIconStyles(state);

  if (state.icon) {
    state.icon.className = ax(iconClassName, state.icon.className);
  }

  return state;
};
