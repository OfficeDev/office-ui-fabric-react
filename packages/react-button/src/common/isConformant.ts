import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: Omit<IsConformantOptions, 'componentPath'>) {
  const defaultOptions = {
    disabledTests: ['has-docblock', 'kebab-aria-attributes', 'as-*'],
    componentPath: module!.parent!.filename.replace('.test', ''),
  };

  baseIsConformant(defaultOptions, testInfo);
}
