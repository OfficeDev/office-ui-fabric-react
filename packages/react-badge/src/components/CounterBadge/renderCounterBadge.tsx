import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { CounterBadgeState } from './CounterBadge.types';
import { counterBadgeShorthandProps } from './useCounterBadge';

export const renderCounterBadge = (state: CounterBadgeState) => {
  const { slots, slotProps } = getSlots(state, counterBadgeShorthandProps);
  let content = '';
  if (state.count !== 0 || (state.count === 0 && state.showZero)) {
    content = state.count > state.overflowCount ? `${state.overflowCount}+` : `${state.count}`;
  }
  return (
    <slots.root {...slotProps.root}>
      {state.iconPosition === 'before' && <slots.icon {...slotProps.icon} />}
      {content}
      {state.iconPosition === 'after' && <slots.icon {...slotProps.icon} />}
    </slots.root>
  );
};
