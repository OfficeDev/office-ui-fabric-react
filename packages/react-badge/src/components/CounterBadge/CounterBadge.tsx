import * as React from 'react';
import { CounterBadgeProps } from './CounterBadge.types';
import { useCounterBadge } from './useCounterBadge';
import { useCounterBadgeStyles } from './useCounterBadgeStyles';
import { renderCounterBadge } from './renderCounterBadge';

/**
 * Define a styled CounterBadge, using the `useCounterBadge` hook.
 * {@docCategory CounterBadge}
 */
export const CounterBadge = React.forwardRef<HTMLElement, CounterBadgeProps>((props, ref) => {
  const state = useCounterBadge(props, ref);
  useCounterBadgeStyles(state);

  return renderCounterBadge(state);
});

CounterBadge.displayName = 'CounterBadge';
