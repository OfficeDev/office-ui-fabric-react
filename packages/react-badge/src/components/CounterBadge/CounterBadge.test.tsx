import * as React from 'react';
import { CounterBadge } from './CounterBadge';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';

describe('CounterBadge', () => {
  isConformant({
    Component: CounterBadge,
    displayName: 'CounterBadge',
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for CounterBadge in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<CounterBadge>Default CounterBadge</CounterBadge>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
