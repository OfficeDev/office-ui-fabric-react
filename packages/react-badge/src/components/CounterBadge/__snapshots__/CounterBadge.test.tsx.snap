// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterBadge renders a default state 1`] = `
<div
  className=" "
>
  0
</div>
`;
